ckan/config/middleware.py

```diff
@@ -248,7 +249,7 @@ def __call__(self, environ, start_response):
         wsgi_input = environ['wsgi.input']
         if isinstance(wsgi_input, FakeCGIBody):
             upload = wsgi_input.vars.get('upload')
-            if upload is not None:
+            if isinstance(upload, FieldStorage):
```

Comment:
Better to use:

    if hasattr(upload.fp, 'close'):
        upload.fp.close()

Reply:
That errors because upload can be None or str.

upload can be None (non-upload regular ckan web site requests), empty string (update resource, create link) or cgi.FieldStorage (regular file uploads). Also note that "if upload:" is false for FieldStorage (previous problem found).

Here is what we want: "if hasattr(upload, 'fp'):".

Then, as you suggest, we don't force type FieldStorage but allow any type which implements "fp". None and str both evaluate as false so no exception is raised.